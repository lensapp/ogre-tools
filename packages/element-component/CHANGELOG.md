# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [20.5.0](https://github.com/lensapp/ogre-tools/compare/v20.4.1...v20.5.0) (2025-02-10)

**Note:** Version bump only for package @lensapp/element-component

### [20.4.1](https://github.com/lensapp/ogre-tools/compare/v20.4.0...v20.4.1) (2025-01-31)

**Note:** Version bump only for package @lensapp/element-component

## [20.4.0](https://github.com/lensapp/ogre-tools/compare/v20.3.0...v20.4.0) (2025-01-30)

**Note:** Version bump only for package @lensapp/element-component

## [20.3.0](https://github.com/lensapp/ogre-tools/compare/v20.2.1...v20.3.0) (2025-01-30)

**Note:** Version bump only for package @lensapp/element-component

### [20.2.1](https://github.com/lensapp/ogre-tools/compare/v20.2.0...v20.2.1) (2025-01-23)

**Note:** Version bump only for package @lensapp/element-component

## [20.2.0](https://github.com/lensapp/ogre-tools/compare/v20.1.0...v20.2.0) (2025-01-22)

### Features

- Introduce first draft of the ElementComponent, ie. reusable React-props ([a7b9b92](https://github.com/lensapp/ogre-tools/commit/a7b9b924fc6468dd9d742bc35117016e0f872996))
