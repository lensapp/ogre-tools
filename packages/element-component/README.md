# Component Element

A way to introduce re-usable props in React.

## Usage

```tsx
const somePlugin = getPlugin((props: ({ $somePluginProp}) => ({ title: props.$somePluginProp })));

const SomeComponentElement = getComponentElement("div", somePlugin);

render(<SomeComponentElement $somePluginProp="some-value" />);

// -> <div title="some-value" />
```
