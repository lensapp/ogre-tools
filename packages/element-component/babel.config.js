module.exports = {
  extends: '@lensapp/infrastructure-babel-for-ts',
};
