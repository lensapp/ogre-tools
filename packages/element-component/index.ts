export {
  getElementComponent,
  ElementComponent,
  TagNames,
} from './src/element-component';

export { getPlugin, Plugin } from './src/plugin/plugin';
