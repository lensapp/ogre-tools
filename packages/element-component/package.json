{
  "name": "@lensapp/element-component",
  "private": false,
  "version": "20.7.0",
  "description": "Re-usable props in React",
  "type": "commonjs",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/lensapp/ogre-tools.git"
  },
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "build"
  ],
  "author": "Ogre Works",
  "license": "MIT",
  "homepage": "https://github.com/lensapp/ogre-tools#readme",
  "peerDependencies": {
    "@lensapp/fp": "^16 || ^17 || ^18 || ^19 || ^20",
    "lodash": "^4.17.21",
    "react": "^18 || ^19",
    "react-dom": "^18 || ^19"
  },
  "devDependencies": {
    "@async-fn/jest": "^1.6.4",
    "@lensapp/discoverable": "^20.7.0",
    "@lensapp/infrastructure-babel-for-ts": "^20.7.0",
    "@lensapp/infrastructure-jest": "^20.7.0",
    "@lensapp/infrastructure-prettier": "^20.7.0",
    "@lensapp/infrastructure-webpack-for-ts": "^20.7.0",
    "@lensapp/ogre-test-utils": "^20.7.0",
    "type-fest": "^4.27.0"
  },
  "scripts": {
    "build": "ogre-tools-build-ts",
    "test": "ogre-tools-test",
    "code-style:verify": "ogre-tools-verify-code-style",
    "code-style:fix": "ogre-tools-fix-code-style"
  }
}
