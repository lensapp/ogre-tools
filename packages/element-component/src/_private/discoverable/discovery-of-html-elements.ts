import type { RenderResult } from '@testing-library/react';
import { fireEvent, prettyDOM } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import { isEmpty } from 'lodash/fp';
import { act } from 'react';
import { flushPromises } from '@lensapp/ogre-test-utils';

type DiscoverySourceTypes = RenderResult | Element | Element[];

export type Actionable = {
  type: (data: string) => Promise<void>;
  click: () => Promise<void>;
  rightClick: () => Promise<void>;
  doubleClick: () => Promise<void>;
  hover: () => Promise<void>;
  unhover: () => Promise<void>;
};
export type Attributable = {
  readonly attributeValue: string | null;
  getAttribute: (name: string) => string | null;
};
export type Discovered = { readonly discovered: Element } & Discover &
  Actionable &
  Attributable;
export type MaybeDiscovered = {
  readonly discovered: Element | null;
} & Discover &
  Actionable;
export type QueryManyDiscovered = {
  readonly discovered: Element[];
  readonly attributeValues: string[];
  getAttributes: (name: string) => (string | null)[];
  getSingleElement: GetSingleElement;
};

export type QuerySingleElement = (
  attributeName: string,
  attributeValue?: string,
) => MaybeDiscovered;
export type GetSingleElement = (
  attributeName: string,
  attributeValue?: string,
) => Discovered;
export type QueryAllElements = (
  attributeName: string,
  attributeValue?: string,
) => QueryManyDiscovered;

export interface Discover {
  querySingleElement: QuerySingleElement;
  queryAllElements: QueryAllElements;
  getSingleElement: GetSingleElement;
}

const getBaseElement = (source: DiscoverySourceTypes) =>
  'baseElement' in source ? source.baseElement : source;

declare global {
  interface DateConstructor {
    /* Jest uses @sinonjs/fake-timers, that add this flag */
    readonly isFake: boolean | undefined;
  }
}

const multiPrettyDom = (getSource: () => DiscoverySourceTypes) =>
  [getBaseElement(getSource())]
    .flat()
    .map(value => prettyDOM(value, Infinity))
    .join('\n');

export function querySingleElement(
  getSource: () => DiscoverySourceTypes,
): QuerySingleElement {
  return (attributeName, attributeValue) => {
    const source = getSource();
    const dataAttribute = `data-${attributeName}-test`;
    const selector = attributeValue
      ? `[${dataAttribute}="${attributeValue}"]`
      : `[${dataAttribute}]`;
    const [discovered = null, ...discoveredExcess] = [getBaseElement(source)]
      .flat()
      .flatMap(elem => [
        elem.matches(selector) ? elem : undefined,
        ...elem.querySelectorAll(selector),
      ])
      .filter(x => x);

    function ensureDiscovered(
      discovered: Element | null,
    ): asserts discovered is Element {
      if (!discovered) {
        throw new Error(
          `Tried to click "${selector}", but it didn't exist.\n\nHTML is:\n\n${multiPrettyDom(
            getSource,
          )}`,
        );
      }
    }

    if (!isEmpty(discoveredExcess)) {
      throw new Error(
        `Tried to discover single "${selector}", but found multiple matches.\n\nHTML is:\n\n${multiPrettyDom(
          getSource,
        )}`,
      );
    }

    const nestedDiscover = discoverFor(() => {
      if (!discovered) {
        throw new Error(
          'Tried to do nested discover using source that does not exist',
        );
      }

      return discovered;
    });
    const isUsingFakeTime = !!global.Date.isFake;
    const configuredUserEvents = userEvent.setup({
      advanceTimers: isUsingFakeTime ? jest.advanceTimersByTimeAsync : () => {},
    });

    const mouseUserEventAction =
      (
        action:
          | 'clear'
          | 'click'
          | 'dblClick'
          | 'hover'
          | 'tripleClick'
          | 'unhover',
      ) =>
      async () => {
        ensureDiscovered(discovered);

        await act(async () => {
          await configuredUserEvents[action](discovered);

          await flushPromises();
        });
      };

    const click = mouseUserEventAction('click');
    const doubleClick = mouseUserEventAction('dblClick');
    const hover = mouseUserEventAction('hover');

    const unhover = async () => {
      // NOTE: This does not use the build in convenience function, as that one does not seem to fire the events consistently
      ensureDiscovered(discovered);
      fireEvent.mouseOut(discovered);
      fireEvent.mouseLeave(discovered);
    };

    const rightClick = async () => {
      ensureDiscovered(discovered);

      await act(async () => {
        await configuredUserEvents.pointer([
          { target: discovered },
          { target: discovered, keys: '[MouseRight]' },
        ]);
      });
    };

    return {
      discovered,
      click,
      rightClick,
      doubleClick,
      hover,
      unhover,
      type: async data => {
        ensureDiscovered(discovered);

        // TODO: remove these two lines once https://github.com/testing-library/user-event/issues/1002 is fixed
        await act(async () => {
          await configuredUserEvents.click(discovered);
        });

        await act(async () => {
          (discovered as HTMLElement).focus();
        });

        await act(async () => {
          await configuredUserEvents.type(discovered, data, {
            skipClick: true,
          });
        });
      },
      ...nestedDiscover,
    };
  };
}

export function queryAllElements(
  getSource: () => DiscoverySourceTypes,
): QueryAllElements {
  return (attributeName, attributeValue) => {
    const source = getSource();
    const dataAttribute = `data-${attributeName}-test`;
    const selector = attributeValue
      ? `[${dataAttribute}="${attributeValue}"]`
      : `[${dataAttribute}]`;
    const results = [getBaseElement(source)]
      .flat()
      .flatMap(elem => [...elem.querySelectorAll(selector)]);

    return {
      discovered: results,
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      attributeValues: results.map(
        result => result.getAttribute(dataAttribute)!,
      ),
      getAttributes: name =>
        results.map(result => result.getAttribute(`data-${name}-test`)),
      getSingleElement: getSingleElement(() => results),
    };
  };
}

export function getSingleElement(
  getSource: () => DiscoverySourceTypes,
): GetSingleElement {
  return (attributeName, attributeValue) => {
    const dataAttribute = `data-${attributeName}-test`;

    const { discovered, ...nestedDiscover } = querySingleElement(getSource)(
      attributeName,
      attributeValue,
    );

    if (!discovered) {
      const html = multiPrettyDom(getSource);

      if (attributeValue) {
        const validValues =
          queryAllElements(getSource)(attributeName).attributeValues;
        const formattedValidValues =
          validValues.map(data => ` - "${data}"`).join('\n') ||
          'NO VALUES FOUND';

        throw new Error(
          [
            `Couldn't find HTML-element with attribute "${dataAttribute}" with value "${attributeValue}".`,
            `Present values are:`,
            formattedValidValues,
            `HTML is:`,
            html,
          ].join('\n\n'),
        );
      }

      throw new Error(
        `Couldn't find HTML-element with attribute "${dataAttribute}"\n\nHTML is:\n\n${html}`,
      );
    }

    return {
      discovered,
      ...nestedDiscover,
      attributeValue: discovered.getAttribute(dataAttribute),
      getAttribute: name => discovered.getAttribute(`data-${name}-test`),
    };
  };
}

export function discoverFor(getSource: () => DiscoverySourceTypes): Discover {
  return {
    querySingleElement: querySingleElement(getSource),
    queryAllElements: queryAllElements(getSource),
    getSingleElement: getSingleElement(getSource),
  };
}
