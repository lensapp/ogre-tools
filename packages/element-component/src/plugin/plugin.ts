import { ArrayValues, UnionToIntersection } from 'type-fest';
import { Simplify } from 'type-fest';

export type Plugin<TInputProps> = (props: TInputProps) => object;

type PluginInputProp<T> = T extends Plugin<infer PluginInput>
  ? PluginInput
  : never;

type MapPluginInputProp<T extends readonly Plugin<any>[]> = {
  [K in keyof T]: PluginInputProp<T[K]>;
};

export type PropsFromPluginTuple<PluginTuple extends readonly Plugin<any>[]> =
  Simplify<UnionToIntersection<ArrayValues<MapPluginInputProp<PluginTuple>>>>;

type PluginWithDependencies<
  TInputProps,
  TDependencyPluginTuple extends readonly Plugin<any>[],
> = Simplify<TInputProps & PropsFromPluginTuple<TDependencyPluginTuple>>;

export function getPlugin<TInputProps>(
  plugin: Plugin<TInputProps>,
): Plugin<TInputProps>;

export function getPlugin<
  TInputProps,
  TDependencyPluginTuple extends readonly Plugin<any>[],
>(
  plugin: Plugin<PluginWithDependencies<TInputProps, TDependencyPluginTuple>>,
): Plugin<PluginWithDependencies<TInputProps, TDependencyPluginTuple>>;

export function getPlugin(plugin: any) {
  return plugin;
}
