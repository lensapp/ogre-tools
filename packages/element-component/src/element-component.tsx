import * as React from 'react';
import { forwardRef, JSX, MutableRefObject } from 'react';
import { Plugin, PropsFromPluginTuple } from './plugin/plugin';
import { withMergeOutputOverInput } from './_private/with-merge-output-over-input';
import { fastPipeline } from './_private/fast-pipeline';
import { withMappedRef } from './_private/with-mapped-ref';

export type TagNames = keyof JSX.IntrinsicElements;

export type ElementComponent<
  TTagName extends TagNames,
  TProps = {},
> = React.ComponentType<TProps & JSX.IntrinsicElements[TTagName]>;

export function getElementComponent<TTagName extends TagNames>(
  tagName: TTagName,
): ElementComponent<TTagName>;

export function getElementComponent<
  TTagName extends TagNames,
  PluginTuple extends readonly Plugin<any>[],
>(
  tagName: TTagName,
  ...plugins: PluginTuple
): ElementComponent<TTagName, PropsFromPluginTuple<PluginTuple>>;

export function getElementComponent<TagName extends TagNames>(
  tagName: TagName,
  ...plugins: any[]
) {
  const processedPlugins = plugins
    .map(withMappedRef)
    .map(withMergeOutputOverInput);

  return forwardRef((unprocessedProps, ref) => {
    const { $$ref = [], ...processedProps }: any = fastPipeline(
      unprocessedProps,
      ...processedPlugins,
    ) as JSX.IntrinsicElements[TagName];
    const callbackRef = React.useCallback(
      (node: HTMLElement) => [ref, ...$$ref].forEach(handleRefFor(node)),
      $$ref,
    );
    const TagName = tagName as React.ElementType;

    return <TagName {...processedProps} ref={callbackRef} />;
  });
}

const handleRefFor =
  (node: HTMLElement) =>
  (
    ref: ((node: HTMLElement) => void) | MutableRefObject<unknown> | undefined,
  ) => {
    if (typeof ref === 'function') {
      ref(node);
    } else if (ref) {
      ref.current = node;
    }
  };
