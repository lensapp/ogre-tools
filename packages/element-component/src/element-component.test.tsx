import * as React from 'react';
import { getElementComponent } from './element-component';
import { RenderResult } from '@testing-library/react';
import { render } from '@testing-library/react';

import { Discover, discoverFor } from '@lensapp/discoverable';

import { getPlugin, Plugin } from './plugin/plugin';
import { useEffect, useRef } from 'react';

describe('element', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error');
  });

  afterEach(() => {
    expect(console.error).not.toHaveBeenCalled();
  });

  describe('given no plugins, when rendered with native HTML-props', () => {
    let rendered: RenderResult;

    beforeEach(() => {
      const Div = getElementComponent('div');

      rendered = render(<Div data-some-native-html-prop-test />);
    });

    it('renders', () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    it('renders the native HTML-prop', () => {
      const discover = discoverFor(() => rendered);

      expect(() => {
        discover.getSingleElement('some-native-html-prop');
      }).not.toThrow();
    });
  });

  describe('given plugin', () => {
    let somePlugin: Plugin<{ $somePluginProp: string }>;

    beforeEach(() => {
      somePlugin = getPlugin(({ $somePluginProp }) => ({
        className: $somePluginProp,
        $somePluginProp: undefined,
      }));
    });

    describe('given plugin and plugin-specific props, when rendered', () => {
      let rendered: RenderResult;

      beforeEach(() => {
        const Div = getElementComponent('div', somePlugin);

        rendered = render(
          <Div
            data-some-element-test
            $somePluginProp="some-plugin-prop-value"
          />,
        );
      });

      it('renders', () => {
        expect(rendered.baseElement).toMatchSnapshot();
      });

      it('is influenced by the plugin', () => {
        const discover = discoverFor(() => rendered);

        const element = discover.getSingleElement('some-element');

        // @ts-ignore
        expect(element.discovered).toHaveClass('some-plugin-prop-value');
      });
    });
  });

  describe('given a plugin depending on another, when rendered', () => {
    let rendered: RenderResult;

    beforeEach(() => {
      const somePlugin1 = getPlugin<{ $somePlugin1Input?: string }>(props => ({
        title: `plugin1(${props.$somePlugin1Input})`,
        $somePlugin1Input: undefined,
      }));

      const somePlugin2 = getPlugin<
        { $somePlugin2Input?: string },
        [typeof somePlugin1]
      >(props => ({
        $somePlugin1Input: `plugin2(${props.$somePlugin2Input})`,
        $somePlugin2Input: undefined,
      }));

      const Div = getElementComponent('div', somePlugin2, somePlugin1);

      rendered = render(
        <Div
          data-some-element-test
          $somePlugin2Input="some-plugin-prop-value"
        />,
      );
    });

    it('renders', () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    it('renders collaborated outcome of plugins', () => {
      const discover = discoverFor(() => rendered);

      const element = discover.getSingleElement('some-element');

      // @ts-ignore
      expect(element.discovered).toHaveAttribute(
        'title',
        'plugin1(plugin2(some-plugin-prop-value))',
      );
    });
  });

  it('given hook ref, when rendered, works', () => {
    const Div = getElementComponent('div');

    const TestComponent = () => {
      const testRef = React.useRef(null);

      useEffect(() => {
        testRef.current.classList.add('some-class-from-ref');
      });

      return <Div ref={testRef} data-some-element-test />;
    };

    const rendered = render(<TestComponent />);

    const discover = discoverFor(() => rendered);

    expect(
      discover
        .getSingleElement('some-element')
        .discovered.classList.contains('some-class-from-ref'),
    ).toBe(true);
  });

  it('given functional ref, when rendered, works', () => {
    const Div = getElementComponent('div');

    const rendered = render(
      <Div
        ref={node => {
          node?.classList.add('some-class-from-ref');
        }}
        data-some-element-test
      />,
    );

    const discover = discoverFor(() => rendered);

    expect(
      discover
        .getSingleElement('some-element')
        .discovered.classList.contains('some-class-from-ref'),
    ).toBe(true);
  });

  describe('given multiple plugins with ref', () => {
    let rendered: RenderResult;
    let discover: Discover;

    beforeEach(() => {
      const somePlugin1 = getPlugin<{ $somePlugin1Input?: boolean }>(
        (props: any): any => {
          const ref = useRef(null);

          useEffect(() => {
            ref.current.classList.add('some-class-from-hook-ref');
          }, []);

          return {
            $somePlugin1Input: undefined,
            ref,
          };
        },
      );

      const somePlugin2 = getPlugin<{ $somePlugin2Input?: boolean }>(
        (props): any => ({
          $somePlugin2Input: undefined,

          ref: node => {
            node?.classList.add('some-class-from-functional-ref');
          },
        }),
      );

      const Div = getElementComponent('div', somePlugin2, somePlugin1);

      rendered = render(
        <Div data-some-element-test $somePlugin1Input $somePlugin2Input />,
      );

      discover = discoverFor(() => rendered);
    });

    it('functional refs work', () => {
      expect(
        discover
          .getSingleElement('some-element')
          .discovered.classList.contains('some-class-from-functional-ref'),
      ).toBe(true);
    });

    it('non functional refs work', () => {
      expect(
        discover
          .getSingleElement('some-element')
          .discovered.classList.contains('some-class-from-hook-ref'),
      ).toBe(true);
    });
  });

  describe('given plugins that override output of previous plugin, when rendered', () => {
    let discover: Discover;
    let rendered: RenderResult;

    beforeEach(() => {
      const somePlugin1 = getPlugin(() => ({
        title: 'irrelevant',
      }));

      const somePlugin2 = getPlugin(() => ({
        title: 'some-plugin-2-value',
      }));

      const Div = getElementComponent('div', somePlugin1, somePlugin2);

      rendered = render(<Div data-some-element-test />);

      discover = discoverFor(() => rendered);
    });

    it('renders only the influence of last plugin', () => {
      const element = discover.getSingleElement('some-element').discovered;

      // @ts-ignore
      expect(element).toHaveAttribute('title', 'some-plugin-2-value');
    });

    it('renders', () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });
  });

  describe('given plugins that contribute to same prop, when rendered', () => {
    let discover: Discover;
    let rendered: RenderResult;

    beforeEach(() => {
      const somePlugin1 = getPlugin(({ title }) => ({
        title: `plugin1(${title})`,
      }));

      const somePlugin2 = getPlugin(({ title }) => ({
        title: `plugin2(${title})`,
      }));

      const Div = getElementComponent('div', somePlugin1, somePlugin2);

      rendered = render(<Div title="some-prop-value" data-some-element-test />);

      discover = discoverFor(() => rendered);
    });

    it('renders all contributions in order', () => {
      const element = discover.getSingleElement('some-element').discovered;

      // @ts-ignore
      expect(element).toHaveAttribute(
        'title',
        'plugin2(plugin1(some-prop-value))',
      );
    });

    it('renders', () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });
  });

  it('given wrong type for a plugin-prop value, when rendering, typing is not ok', () => {
    const somePlugin = getPlugin<{ $somePluginProp: number }>(() => ({
      $somePluginProp: undefined,
    }));

    const Div = getElementComponent('div', somePlugin);

    // @ts-expect-error
    void render(<Div $somePluginProp="some-not-a-number" />);
  });

  it('given plugin not cleaning up for itself, typing is not ok', () => {
    // @ts-expect-error
    void getPlugin<{ $somePluginProp?: number }>(() => ({
      // Notice: this is "not cleaned up":
      // $somePluginProp: undefined,
    }));
  });

  it("given a plugin-prop which doesn't clean up after its input prop, when rendering, logs error", () => {
    console.error
      // @ts-ignore
      .mockImplementationOnce(() => {})
      .mockImplementationOnce(() => {});

    // @ts-expect-error
    const somePlugin = getPlugin<{ $somePluginProp: string }>(() => ({}));

    const Div = getElementComponent('div', somePlugin);

    void render(<Div $somePluginProp="some-string" />);

    expect(console.error).toHaveBeenCalledTimes(2);

    // @ts-ignore
    console.error.mockClear();
  });

  it('given wrong type for a native prop value, when rendering, typing is not ok', () => {
    const Div = getElementComponent('div');

    // @ts-expect-error
    void render(<Div href="some-string" />);
  });
});
