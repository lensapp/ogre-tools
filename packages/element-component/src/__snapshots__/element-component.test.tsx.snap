// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`element given a plugin depending on another, when rendered renders 1`] = `
<body>
  <div>
    <div
      data-some-element-test="true"
      title="plugin1(plugin2(some-plugin-prop-value))"
    />
  </div>
</body>
`;

exports[`element given no plugins, when rendered with native HTML-props renders 1`] = `
<body>
  <div>
    <div
      data-some-native-html-prop-test="true"
    />
  </div>
</body>
`;

exports[`element given plugin given plugin and plugin-specific props, when rendered renders 1`] = `
<body>
  <div>
    <div
      class="some-plugin-prop-value"
      data-some-element-test="true"
    />
  </div>
</body>
`;

exports[`element given plugins that contribute to same prop, when rendered renders 1`] = `
<body>
  <div>
    <div
      data-some-element-test="true"
      title="plugin2(plugin1(some-prop-value))"
    />
  </div>
</body>
`;

exports[`element given plugins that override output of previous plugin, when rendered renders 1`] = `
<body>
  <div>
    <div
      data-some-element-test="true"
      title="some-plugin-2-value"
    />
  </div>
</body>
`;

exports[`element given ref, when rendered works 1`] = `
<body>
  <div>
    <div
      data-some-native-html-prop-test="true"
    />
  </div>
</body>
`;
